graphrag/index/flows/create_final_embeddings.py

```diff
@@ -0,0 +1,211 @@
+# Copyright (c) 2024 Microsoft Corporation.
+# Licensed under the MIT License
+
+"""All the steps to transform the text units."""
+
+import logging
+from typing import cast
+
+import pandas as pd
+from datashaper import (
+    VerbCallbacks,
+)
+
+from graphrag.index.cache import PipelineCache
+from graphrag.index.config.embeddings import (
+    community_full_content_embedding,
+    community_summary_embedding,
+    community_title_embedding,
+    document_raw_content_embedding,
+    entity_description_embedding,
+    entity_name_embedding,
+    relationship_description_embedding,
+    text_unit_text_embedding,
+)
+from graphrag.index.operations.embed_text import embed_text
+from graphrag.index.operations.snapshot import snapshot
+from graphrag.index.storage import PipelineStorage
+
+log = logging.getLogger(__name__)
+
+
+async def create_final_embeddings(
+    final_documents: pd.DataFrame,
+    final_relationships: pd.DataFrame,
+    final_text_units: pd.DataFrame,
+    final_entities: pd.DataFrame,
+    final_community_reports: pd.DataFrame,
+    callbacks: VerbCallbacks,
+    cache: PipelineCache,
+    storage: PipelineStorage,
+    embedded_fields: set[str],
+    base_text_embed: dict | None = None,
+) -> pd.DataFrame:
+    """All the steps to generate all embeddings."""
+    documents_embeddings = final_documents[["id", "raw_content"]].copy()
+    relationships_embeddings = final_relationships[["id", "description"]].copy()
+    text_units_embeddings = final_text_units[["id", "text"]].copy()
+
+    entities_embeddings = final_entities[["id", "name", "description"]].copy()
+    entities_embeddings["name_description"] = (
+        entities_embeddings["name"] + ":" + entities_embeddings["description"]
+    )
+
+    community_reports_embeddings = final_community_reports[
+        ["id", "full_content", "summary", "title"]
+    ].copy()
+
+    if base_text_embed:
+        log.info("Creating embeddings")
+        # Embed raw_content in documents table
+        if document_raw_content_embedding in embedded_fields and isinstance(
+            documents_embeddings, pd.DataFrame
+        ):
+            await run_embeddings(
+                data=documents_embeddings,
+                column_to_embed="raw_content",
+                filename="create_final_documents_raw_content_embeddings",
+                callbacks=callbacks,
+                cache=cache,
+                storage=storage,
+                base_text_embed=base_text_embed,
+                message="Documents embeddings (raw content)",
+            )
+
+        # Embed description in relationships table
+        if relationship_description_embedding in embedded_fields and isinstance(
```

Comment:
I think we don't need to run the instance check. These params are already data frames, and there's slicing done above. If not data frames, at this point code would've already crashed.

Reply:
yes I agree but the ruff check requested the validation to not throw errors.

Reply:
Same as below, if we use slicing that ensures df retrieval we can avoid type checks and casting